File: apps/web/src/lib/trading.ts
export type Account = {
  id: string;
  name: string;
  kind: "paper" | "live_hl_testnet" | "live_hl_main";
  owner_id: string | null;
  starting_capital: string;
  is_active: boolean;
  meta: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type PortfolioPosition = {
  account_id: string;
  symbol: string;
  qty: string;
  avg_price: string;
  realized_pnl: string;
  last_price: string;
  updated_at: string;
};

export type PortfolioFill = {
  id: string;
  order_id: string;
  account_id: string;
  symbol: string;
  side: "buy" | "sell";
  qty: string;
  price: string;
  fee: string;
  pnl: string;
  ts: string;
};

export type PortfolioSummary = {
  account: Account;
  cash: string;
  equity: string;
  total_return: number;
  realized_pnl: string;
  unrealized_pnl: string;
  open_positions: number;
  positions: PortfolioPosition[];
  recent_fills: PortfolioFill[];
};

export function num(s: string | number | null | undefined): number {
  if (s === null || s === undefined) return 0;
  return typeof s === "number" ? s : Number(s);
}

File: apps/api/src/maelstrom_api/schemas/trading.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class AccountOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    kind: str
    owner_id: uuid.UUID | None = None
    starting_capital: Decimal
    is_active: bool
    killed: bool = False
    daily_loss_limit_pct: Decimal | None = None
    meta: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class AccountCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    kind: str = Field(default="paper")
    starting_capital: Decimal = Field(default=Decimal("10000"))
    daily_loss_limit_pct: Decimal | None = None
    meta: dict[str, Any] = Field(default_factory=dict)


class AccountUpdate(BaseModel):
    is_active: bool | None = None
    daily_loss_limit_pct: Decimal | None = None
    meta: dict[str, Any] | None = None


class PositionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    account_id: uuid.UUID
    symbol: str
    qty: Decimal
    avg_price: Decimal
    realized_pnl: Decimal
    last_price: Decimal
    updated_at: datetime


class OrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    account_id: uuid.UUID
    live_strategy_id: uuid.UUID | None = None
    source: str
    symbol: str
    side: str
    qty: Decimal
    price: Decimal | None = None
    order_type: str
    status: str
    filled_qty: Decimal
    avg_fill_price: Decimal
    error: str | None = None
    reason: str | None = None
    exchange_order_id: str | None = None
    submitted_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime


class FillOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order_id: uuid.UUID
    account_id: uuid.UUID
    symbol: str
    side: str
    qty: Decimal
    price: Decimal
    fee: Decimal
    pnl: Decimal
    ts: datetime


class PortfolioSummary(BaseModel):
    account: AccountOut
    cash: Decimal
    equity: Decimal
    total_return: float
    realized_pnl: Decimal
    unrealized_pnl: Decimal
    open_positions: int
    positions: list[PositionOut]
    recent_fills: list[FillOut]


class EquityPointOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    ts: datetime
    equity: Decimal
    cash: Decimal
